### ISSUE TITLE
`commandline` needs serious cleanup

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#7375]
While the overall fish docs are fairly good, there are a few sections where I've never been able to reconcile what fish does with what the docs say it *should* do, and `commandline` is the biggest offender here. Depending on who wrote the documentation, it can be viewed as either a spec dictating how individual fish commands should behave or it can be viewed as an interpretation of existing fish behavior, codified in text.

Depending on which view you take, either the fish documentation for `commandline` needs to be completely redone or the builtin needs some serious refactoring and reorganization to line up with what the documentation suggests is the correct behavior for the various command line arguments it accepts.

When I first started writing completions, I ended up writing a lot of extra code to parse the command line because docs didn't make it apparent that certain functionality was already possible simply by passing certain flags to the `commandline` builtin. At the same time, certain behavior the docs suggest is possible doesn't seem to actually be present.

My biggest issue is with the `-c`/`--cut-at-cursor` option. It is explicitly stated that its use has the effect of truncating the contents of the commandline buffer to the location of the cursor:

> only print selection up until the current cursor position

i.e. with `\cp` bound to the following function:

```fish
function debug
    printf '\n%s' (commandline -co)
    printf '\n'
    commandline -f repaint
end
```

I would expect the following behavior when ctrl+p is pressed with the contents of the buffer as follows and the cursor at the underline position:

```
$ hello wor̲ld friend
hello
wo
$ hello wor̲ld friend
```

while the *actual* behavior is as follows:

```
$ hello wor̲ld friend
hello
$ hello wor̲ld friend
```

Now as a fish developer and someone that's now been exposed to and used `commandline` a fair amount, if you asked me what `commandline -c` does, I would describe its current behavior as follows:

> The `-c`/`--cut-at-cursor` option prints the output of the commandline buffer up to and including the token/argument that directly precedes the token/argument that the cursor is currently in

(which is probably extremely confusing to read and almost certainly can be phrased in an easier to understand way)

This distinction has a fair amount of repercussions:

* it means that `printf '%s\n' (commandline -co)[-1]` can be used to get the token directly preceding the token being completed without resorting to needing to decide between index -1 and index -2 based off of whether or not the command line buffer ends with whitespace (and that's without dealing with cases like `$ hello "world `), but
* it means that one has to resort to inspecting the full contents of the command line buffer to determine what the *actual* current token being completed is,  leading to ugly code [like this](https://github.com/fish-shell/fish-shell/commit/010db9e7be48f2a80e92fed8ab45c5f91edfa9c5#diff-a6f0e9b903f341e4558018feec56576aR197-R280) to introspect the current token and to generate output that depends on it, (edit: `commandline -ct` can be used to get the current token truncated to the current cursor position (so no way to get the current token not truncated?), which makes the behavior of `-c` dependent on the other switches it's being used in conjunction with)
* I could be wrong, but I think it means it's actually impossible to generate correct completions mid-word if introspection of the current argument is required, because the options are "contents with the current token and all subsequent tokens completely excluded" and "contents with all tokens included," meaning if completions are generated dependent on the current token, attempting to generate a completion for `$ hello wor̲ld friend` at the marked location will fail because it will not disregard `rld friend`

**Comment by faho** [comment]
Is this a duplicate of #3934.

**Comment by faho** [comment]
>it means that one has to resort to inspecting the full contents of the command line buffer to determine what the actual current token being completed is

>(edit: commandline -ct can be used to get the current token truncated to the current cursor position (so no way to get the current token not truncated?),

The current token can be gotten with `commandline --current-token`/`commandline -t`.  You don't need to give cut-at-cursor if you don't want.

>I could be wrong, but I think it means it's actually impossible to generate correct completions mid-word if introspection of the current argument is required, because the options are "contents with the current token and all subsequent tokens completely excluded" and "contents with all tokens included,

`commandline --current-process --tokenize --cut-at-cursor; commandline --current-token` (short: `commandline -poc; commandlie -t`) should give you what you want.

I get that it's a bit awkward, but I'm struggling to find something that's

1. better
2. backwards-compatible

Because typically you *don't* want to react to the current token in a completion, especially not the incomplete version up to the cursor.

E.g. if you have `git sta<TAB>tus`, your script should just check all the git subcommands and print all of them, and then fish goes and matches `status` and completes the token.

If you react to the current token, you typically break fuzzy matching. There are cases where it's necessary (e.g. when there are too many possible completions), but the default should be to just check the tokens *before* the cursor.

But anyway, I still think this is the same thing as #3934, so I'm closing that so we can restart the discussion here.

**Comment by mqudsi** [comment]
Yeah, I was going to say I think this PR describes the overarching issue that #3934 is a symptom of.

At the very least, if we accept the current behavior as "as good as it's going to get" then the documentation should be changed to provide actually usable descriptions for the behavior of the flags. I think it's at least important to highlight that the behavior of `-c` is dependent on other switches passed to `commandline`.

> If you react to the current token, you typically break fuzzy matching. There are cases where it's necessary (e.g. when there are too many possible completions), but the default should be to just check the tokens before the cursor.

I agree about the defaults, my issue is solely with the awkwardness of getting desired non-default behavior out of `commandline`. I'm aware of the benefits of generating all completions and leaving it to the shell to determine which to show based on its fuzzy logic, the only exceptions I've encountered have been a) too many completions (as you said), primarily only when dealing with things that dump a database such as `apt` or `npm`, and b) additive completions that include multiple "payloads" within a single token (per fish's definition of what a token is) delimited by some arbitrary or convoluted construct, such as the completions for `ffmpeg`'s filter options (sometimes delimited by `:`, sometimes by `=`, sometimes by `,`).

(There's an argument to be made for enhancing the `commandline` logic with the ability to customize the tokenization of the current fish token to support such constructs, but real world usages are few and far enough in between that I wouldn't waste the effort there.)

I would keep the current switches as-is for backwards compatibility reasons, but we could perhaps introduce subcommands to clear up the current mess, e.g. `commandline [append|replace|insert|print|tokenize]` (or maybe tokenize should stay a `-t` to `commandline print`, but this is just all off the top of my head), and when a subcommand is specified, a new option set is used.

I think `-c` has to go, because it sometimes actually means "cut at cursor" (which is what it is short for) but at other times it just means "cut at last completed token". I would bubble it up and make it a full-fledged truncation option, all the options being

* default, aka to end of previous token: tokens prior to the one the cursor is currently in, that have been completely entered and tokenized, e.g. followed by trailing whitespace,
* to cursor, actually being what it says it does: from the start of the commandline up to but not including the current insertion position, including the token the cursor is currently in,
* to end of current token: unlikely to be used, but prints the commandline up to and including the end of the token that the cursor is currently in,
* in its entirety: the entire contents of the command line, not truncated

(I just remembered another issue with the `commandline` docs: in several places it says "including any completions," which I always understood to mean "including the ghost completion from history," which could just be me but imho we could use more precise language than that.)

The truncation options above specify where the output ends, they can optionally be combined with an option for where the output starts:

* default or start of current command, e.g. `$ echo (command hello <TAB>)` starts at `hello`
* actual start of prompt, e.g. `$ echo (hello <TAB>` starts at `echo`
* start of current expression, e.g. `echo (command cat foo<TAB>` starts at `command`
* start of arguments for current command, e.g. `tail -n 20 foo.<TAB>` starts at `-n`
* start of previous token, e.g. '$ echo (hello world <TAB>` starts at `world`
* start of current token, e.g. `$ echo (hello worl<TAB>` starts at `worl`

and a combination of selected start and end position may result in a null result set, e.g. default truncation at end of previous token

**Comment by faho** [comment]
I've decided to add something to the docs, for now. I couldn't find "including any completions" in the docs anywhere so I think that was removed already?

If someone does want to work on this, we would need more of a write-up of

1. What the problems are - is it that `-c` is weird with `-o`?
2. A plan for fixing it, ideally backwards-compatibly, or with an acceptable transition (not saying that we should never change anything here, but it would have to be manageable)

That would be easiest to manage in a new issue.
